contracts/Spt.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+abstract contract Spt {
+    // because maxElements = 2**depth, so 256 will overflow
+    uint constant MAX_DEPTH = 255;
+    bytes32 constant internal EMPTY_LEAF = 0x00;
+
+    mapping(uint256 => bytes32) public cacheEmptyValues;
+
+    uint public depth;
+    uint public maxElements;
+
+    // tree level => index inside level => element hash
+    mapping(uint256 => mapping(uint256 => bytes32)) public tree;
+    mapping(uint256 => bytes) public elementData; 
+
+    constructor(uint _depth) {
+        require(_depth > 0, "Depth must be non-zero");
+        require(_depth <= MAX_DEPTH, "Overflow protection");
+        setupDepth(_depth);
+        calculateEmptyLeafHash(0, _depth);
+    }
+
+    function setupDepth(uint _depth) private {
+        depth = _depth;
+        // 1<<depth == 2**depth
+        maxElements = 1<<depth;
+    }
+
+    function increaseDepth(uint depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint oldDepth = depth;
+        require(depthDifference <= MAX_DEPTH - oldDepth, "Overflow protection");
+        uint newDepth = oldDepth + depthDifference;
+
+        calculateEmptyLeafHash(oldDepth+1, newDepth);
+
+        for (uint level = oldDepth+1; level <= newDepth; level++) {
+            updateLeaf(level, 0);
+        }
+        setupDepth(newDepth);
+    }
+
+    function decreaseDepth(uint depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint oldDepth = depth;
+        require(depthDifference < oldDepth, "Overflow protection");
+        uint newDepth = oldDepth - depthDifference;
+
+        for (uint level = newDepth; level < oldDepth; level++) {
+            require(tree[level][1] == EMPTY_LEAF, "Subtree must be empty");
+        }
```

Comment:
This is quite confusing, but I think that it is actually fine.
Note that, at each level, we check node index 1; I'm not sure which "root hash" you mean (either in the old or new tree) but:
1. In the old tree, the _require_  don't check the top level (and anyway, it doesn't have index 1).
2. In the new tree, we don't check node 0, which is the one that has been modified (by _addElement).

Reply:
I mean that we had an empty tree with height `x`, we added an element with index `0` at height `x` and found the root hash of the tree, and then reduced the size of the tree by 1, but the value in the leaf at height `x-1` has already been calculated, this is clearly not what you expect from the function of decreasing the height of the tree

Reply:
We wrote test for this case

```
def test_insert_and_decrease(public_spt, accounts):

    spt = PublicSha256Spt.deploy(10, {"from": accounts[0]})
    spt._addElement(0, b"lol")

    elhash = sha256(b"lol").digest()
    empty = sha256(b"").digest()
    r = elhash
    for item in range(0, 10):
        r = sha256(r + empty).digest()
        empty = sha256(empty + empty).digest()
    
    assert spt.getRoot() == '0x' + r.hex()
    
    spt._decreaseDepth(9)

    assert spt.getRoot() == '0x' + sha256(sha256(b"lol").digest() + sha256(b"").digest()).hexdigest()
```

Reply:
I may did not understand what you meant by saying that it will be broken: incorrect implementation or that the function would err. My previous comment assumed the latter.
If you were talking about incorrect implementation, then "decreasing the tree" is really just to keep some (left) subtree, and ignore the rest. The subtree is self-contained/consistent, so in particular the root is computed correctly.
All this function does is making sure that no one calls if some leaves that we are going to disregard are non-empty. In other words, that it is ok to decrease the tree as all the other values add no information.